
<div style="border-radius:10px;
            border : #015a2c solid;
            background-color:#D7D9DB;
           font-size:110%;
           letter-spacing:0.5px;
            text-align: center">
​
<center><h1 style="padding: 25px 0px; color:#015a4c; font-weight: bold; font-family: Cursive">
Amit Project for Data Science Module <br><br>💡💡💡💡💡 <br><br> DEPI_BS_DS_R3  </h1></center>
<center><h3 style="padding-bottom: 25px; color:#025b2c; font-weight: bold; font-style:italic; font-family: Cursive">
Using Project: Trip data Ford go bike  </h3></center>     
​
</div>

# Part I - (Dataset Exploration for Ford GoBike)

### worked by (abanob nabil kamal /2025 )

## Introduction

.

> **Let's introduce our dataset:**


======================================

.

**This data set information was produced from `Ford GoBike` System** the information in it talked about individual rides that were made in a bike-sharing system for covering San Francisco Bay area.** 

.




> ## And here are the main exploring questions 



   **==================================================**



- What is the structure of your dataset?


- What are the main interesting features in the dataset?


- What features in the dataset supported my investigation for these interesting features?




.

> ### Also here are the investigative questions classified into three types



   **================================================**
   
   


###  <a href='#011'> Univariate Exploration</a> 

- <a href='#1'> 1 ) . What is the distribution of `User type`?</a>  
- <a href='#2'> 2 ) . What is the distribution for the `bike share users on all Trips`?</a>  
- <a href='#3'> 3 ) . What is the distribution for users' `ages`?</a>  
- <a href='#4'> 4 ) . Do the users' ages have `outlier` values?</a>
- <a href='#5'> 5 ) . What is the distribution for the trip duration in `minutes`? </a>
- <a href='#6'> 6) . What is the distribution for `member_gender`? </a>
- <a href='#7'> 7) . What is the distribution for the trip duration in `Hours`?  </a>




##  <a href='#111'> Bivariate Exploration</a> 


- <a href='#8'> 8 ) . What the `average duration` of trip for the categories of `user type`? </a>
- <a href='#9'> 9 ) . How long is the duration of the trip for `each gender` in `minutes`? </a>
- <a href='#10'> 10 ) . How long is the duration of the trip for each `user_type` in `minutes`?  </a>
- <a href='#11'> 11 ) . How the user distribution depending on their `Age` and `user_type` ?</a>
- <a href='#12'>12 ). How the user distribution depending on their `Age` and `member_gender` ?  </a>


##  <a href='#1111'> Multivariate Exploration</a> 


- <a href='#13'> 13 ) . What is the Correlation between `each` variable in the dataset?</a>  
- <a href='#14'> 14 ) . What is the Correlation between `specific numeric variables` in the dataset?</a>  
- <a href='#15'> 15 ) . What is the  Correlation between `data_stat` variables in the dataset?</a>  


##  <a href='#1100'> Conclusions </a> 





# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Import Libraries</p>

In [20]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
import matplotlib.pyplot as plt
import seaborn as sns
# %matplotlib inline

In [21]:
# Load my dataset using pandas Dataframe
path='fordgobike-tripdataFor201902.csv'

df = pd.read_csv(path)

In [63]:
# here we make overview in high-level for data shape and composition

df.shape

(183416, 16)

In [75]:
unnessery_col = ["start_station_id","end_station_id","bike_id"]
df.drop(unnessery_col,axis=1,inplace=True)

In [24]:
## check each var  datatype 
type = df.dtypes
num = df.nunique()
pd.DataFrame({
    "type" : type,
    "num" : num
}).T

,duration_sec,start_time,end_time,start_station_id,start_station_name,start_station_latitude,start_station_longitude,end_station_id,end_station_name,end_station_latitude,end_station_longitude,bike_id,user_type,member_birth_year,member_gender,bike_share_for_all_trip
type,int64,str,str,float64,str,float64,float64,float64,str,float64,float64,int64,str,float64,str,str
num,4752,35783,35767,329,329,334,335,329,329,335,335,4646,2,75,3,2


In [79]:
cat_col = ["user_type","member_gender","bike_share_for_all_trip"]
num_col = ["start_time","end_time","start_station_name","start_station_latitude","start_station_longitude","end_station_name","end_station_latitude","end_station_longitude","member_birth_year"]

In [83]:
## I display 15 rows to understand the content values more 
df.head(15)

,duration_sec,start_time,end_time,start_station_name,start_station_latitude,start_station_longitude,end_station_name,end_station_latitude,end_station_longitude,user_type,member_birth_year,member_gender,bike_share_for_all_trip
0,52185,32:10.1,01:56.0,Montgomery St BART Station (Market St at 2nd St),37.789625,-122.400811,Commercial St at Montgomery St,37.794231,-122.402923,Customer,1984.0,Male,No
1,42521,53:21.8,42:03.1,The Embarcadero at Steuart St,37.791464,-122.391034,Berry St at 4th St,37.775880,-122.393170,Customer,NaN,NaN,No
2,61854,13:13.2,24:08.1,Market St at Dolores St,37.769305,-122.426826,Powell St BART Station (Market St at 4th St),37.786375,-122.404904,Customer,1972.0,Male,No
3,36490,54:26.0,02:36.8,Grove St at Masonic Ave,37.774836,-122.446546,Central Ave at Fell St,37.773311,-122.444293,Subscriber,1989.0,Other,No
4,1585,54:18.5,20:44.1,Frank H Ogawa Plaza,37.804562,-122.271738,10th Ave at E 15th St,37.792714,-122.248780,Subscriber,1974.0,Male,Yes
5,1793,49:58.6,19:51.8,4th St at Mission Bay Blvd S,37.770407,-122.391198,Broadway at Kearny,37.798014,-122.405950,Subscriber,1959.0,Male,No
6,1147,55:35.1,14:42.6,Palm St at Willow St,37.317298,-121.884995,San Jose Diridon Station,37.329732,-121.901782,Subscriber,1983.0,Female,No
7,1615,41:06.8,08:02.8,Washington St at Kearny St,37.795393,-122.404770,Valencia St at 21st St,37.756708,-122.421025,Subscriber,1989.0,Male,No
8,1570,41:48.8,07:59.7,Washington St at Kearny St,37.795393,-122.404770,Valencia St at 21st St,37.756708,-122.421025,Subscriber,1988.0,Other,No
9,1049,49:47.7,07:17.0,Post St at Kearny St,37.788975,-122.403452,Mission Playground,37.759210,-122.421339,Subscriber,1992.0,Male,No


In [82]:
# descriptive statistics for numeric variables
df[num_col].describe()


,start_station_latitude,start_station_longitude,end_station_latitude,end_station_longitude,member_birth_year
count,183416.000000,183416.000000,183416.000000,183416.000000,175151.000000
mean,37.771221,-122.352661,37.771425,-122.352247,1984.806504
std,0.099586,0.117100,0.099494,0.116677,10.116589
min,37.317298,-122.453705,37.317298,-122.453705,1878.000000
25%,37.770083,-122.412408,37.770407,-122.411726,1980.000000
50%,37.780760,-122.398285,37.781010,-122.398279,1987.000000
75%,37.797280,-122.286533,37.797320,-122.288045,1992.000000
max,37.880222,-121.874119,37.880222,-121.874119,2001.000000




> ### What is the structure of your dataset?

The dataset contains 16 columns and 183,416 rows.
 
 
- Most variables are categorical and numerical in nature , and here are **dtypes summary :**

        - str (7)
        - float (7)
        - int (2)
        



> ### What are the main interesting features in the dataset?
Trip duration
start_time 
end_time 
start_station_name 
end_station_name 
user_type 
member_birth_year 
member_gender 
bike_share_for_all_trip 
 **We got three main feature areas which are :**

 
- duration_sec, start_time, end_time
 
 
 
- start_station, end_station
 
  
 
  
- user_type, member_gender, member_birth_year




> ### What features in the dataset supported my investigation for these interesting features?



- Trip Information: duration_sec, start_time, end_time



- start_station_name, end_station_name, start_station_latitude, start_station_longitude, end_station_latitude, end_station_longitude



- user_type, member_gender, member_birth_year



- bike_id




- 




##  More Exploring ideas to begin my investigation

In [70]:
#This code for describing statistical data ideas in a more organized and readable table
describe = df[num_col].describe()

pd.DataFrame(describe) 


,start_station_latitude,start_station_longitude,end_station_latitude,end_station_longitude,member_birth_year
count,183416.000000,183416.000000,183416.000000,183416.000000,175151.000000
mean,37.771221,-122.352661,37.771425,-122.352247,1984.806504
std,0.099586,0.117100,0.099494,0.116677,10.116589
min,37.317298,-122.453705,37.317298,-122.453705,1878.000000
25%,37.770083,-122.412408,37.770407,-122.411726,1980.000000
50%,37.780760,-122.398285,37.781010,-122.398279,1987.000000
75%,37.797280,-122.286533,37.797320,-122.288045,1992.000000
max,37.880222,-121.874119,37.880222,-121.874119,2001.000000


In [64]:
# display the information about each column

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 183416 entries, 0 to 183415
Data columns (total 16 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             183416 non-null  int64  
 1   start_time               183416 non-null  str    
 2   end_time                 183416 non-null  str    
 3   start_station_id         183219 non-null  float64
 4   start_station_name       183219 non-null  str    
 5   start_station_latitude   183416 non-null  float64
 6   start_station_longitude  183416 non-null  float64
 7   end_station_id           183219 non-null  float64
 8   end_station_name         183219 non-null  str    
 9   end_station_latitude     183416 non-null  float64
 10  end_station_longitude    183416 non-null  float64
 11  bike_id                  183416 non-null  int64  
 12  user_type                183416 non-null  str    
 13  member_birth_year        175151 non-null  float64
 14  member_gender  

## <mark>The above result show some missing values exist in some columns So let's display them </mark>

In [71]:
#yOUR CODES
df.isnull().sum()

duration_sec                  0
start_time                    0
end_time                      0
start_station_id            197
start_station_name          197
start_station_latitude        0
start_station_longitude       0
end_station_id              197
end_station_name            197
end_station_latitude          0
end_station_longitude         0
bike_id                       0
user_type                     0
member_birth_year          8265
member_gender              8265
bike_share_for_all_trip       0
dtype: int64



> ## <mark>The result above shows many missing data </mark>



###  To fix this issue You will fill these values with median for numeric and mode for categorical

In [81]:
# trying to fill missing values in my datset using for loop 
df[num_col] = df[num_col].astype("int")
df[cat_col] = df[cat_col].astype("category")
print(num_col)
print(cat_col)
# for col in num_col :
#     df[col]= df[col].fillna(df[col].median())
    
# for col in cat_col :
#     df[col]= df[col].fillna(df[col].mode())
    

ValueError: invalid literal for int() with base 10: '32:10.1'

In [33]:
#check the Nulls 


### **And with a proper category for other (member gender)**

In [34]:
##  still got a missing value in the member_gender, so let's explore this column more 

# print(gbf.member_gender.head(15))

In [35]:
# fill  dataset with missing values.
    


In [36]:
# print(gbf.member_gender.head(15))

In [37]:
## check my result above 



> ## <mark>The result above shows that we deal with all missing data </mark>

### Now let's check duplicate value

In [38]:
##  check duplicate in dataset



> ##If Not significant duplicated , so won't clean it 


### Now let's check the category columns to make more cleaning dataset 

In [39]:
### check the category columns that included object


 > ## From all previous results above I found  :
 
 
 
 **=========================================================**
 
 

> 1 ) **unique Value** as in ((`member_gender`)) : Not significant , so won't clean it



> 2 ) **Incorrect data type** as in (( `user_type` &&	`member_gender`)) :  significant change , so I need to clean it.


> 3 ) **Good factor** for marketing work as in ((`member_birth_year`)) : so I need to clean it.

In [40]:
##  step  2 ) Incorrect data type as in (( `user_type` &&	`member_gender`)) :  significant change , so  need to clean it.



In [41]:
## check your work 



In [42]:
## step 3 ) Good factor for marketing work as in ((`member_birth_year`)) : so need to clean it.
## To know the target audience's age and serve on them

# create a column for users' age depending on the current year



In [43]:
## check your work 



In [44]:
## change the new column age data type from float to int



In [45]:
## create a function to know the average users ages


In [46]:
## another important factor to offer  marketing services is the trip duration time 

## need to change the time from second to minute to be more usable




In [47]:
## I need to change the time from second to Houre  to be more usable



In [48]:
## check your work 



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Univariate Exploration</p>
# <a id="011"> Univariate Exploration</a>



> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.




# <a id="1"> 1 ) . What is the distribution for User type?</a>

In [49]:
## I will import seaborn library to support me in making visualizations 

import seaborn as sns
# plt.figure(figsize = (10,5), dpi = 200)

# plot = sns.countplot(data=gbf, x='bike_share_for_all_trip')

# plt.xlabel('User distribution')
# plt.ylabel('Frequency')
# plt.title("Represent the user's distribution depending on being a part of the Trips Program")

# plt.show()

> ### <mark>Your analsys and exploration here </mark>

# <a id="2"> 2 ) What is the distribution for the bike share users on all Trips?</a>

> ### <mark>Your analsys and exploration here </mark>

# <a id="3"> 3 ) What is the distribution for users' ages? </a>

> ### <mark>Your analsys and exploration here </mark>


# <a id="4">  4 ) . Do the users' ages have outlier values?</a>


> ### <mark>Your analsys and exploration here </mark>

# <a id="5">  5 ) . What is the distribution for the trip duration in minutes? </a>

In [50]:
## Review the statistic for trip duration in minutes



In [51]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>


# <a id="6"> 6 ) . What is the distribution for member_gender?</a>

In [52]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="7">    7 )  . What is the distribution for the trip duration in Hours?  </a>

In [53]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

### The variables distribution:


> The distribution of duration present that most of the trips were less than 30 minutes ( we can notice that from the minutes and the hour charts )


> And the Age distribution is ( left Skewed ) as most users age are  22 - 39


> Most users are subscribers, not Customers and  Males as shown in the charts above.



### Were there any interesting and outliers 


- We got Age outliers for for more than 100 .



- The distribution of duration was strange where the minutes and the hour charts were 30 minutes or less.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Bivariate Exploration</p>
# <a id="111"> Bivariate Exploration</a>

Now let's investigate relationships between pairs of variables in our data.

# <a id="8"> 8) . What the average duration of trip for the categories of user type?</a>

In [54]:
# Distribution of Age by user types



# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="9">    9 ) . How long is the duration of the trip for each gender in minutes?  </a>

In [55]:
# Make visualization to easy get insight

# ------------------------Duration distribution by member gender



# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="10 "> 10 ) . How long is the duration of the trip for each user_type in minutes?  </a>

In [56]:

# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

# <a id="11">  11 ) . How the user distribution depending on their Age and user_type ? </a>

In [57]:

# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

#  <a id="12"> 12 ). How the user distribution depending on their Age and member_gender ?</a>

In [58]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> ...................


### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> ....................

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Multivariate Exploration</p>
# <a id="1111"> Multivariate Exploration </a>

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

In [59]:
## check your work 



# <a id="13">  13 ) . What is the  Correlation between each variable in the dataset?</a>

In [60]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes/hours`)



# <a id="14"> 14 )  .  What is the  Correlation between `specific numeric variables` in the dataset?</a>


In [61]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- 




- 




# <a id="15">  15 ) What is the  Correlation between data_stat variables in the dataset?</a>


In [62]:

"""""This time I will use heatmap  to explain how Correlation between data_stat variables

in the dataset in more clear and specific percentage"""

# Make visualization to easy get insight



'""This time I will use heatmap  to explain how Correlation between data_stat variables\n\nin the dataset in more clear and specific percentage'

> ### <mark>The above charts show how Correlation between specific numeric variables in the dataset using heatmap plots  </mark>



**Actually the `heat map`  display the above ideas in clear percentages**



- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- There is a `negative  Correlation` between  member_birth_year  and age which normal.




- There is `no significant Correlation` for any variable with `bike_id `



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Conclusions</p>
# <a id="1100"> Conclusions </a>

#example
- Most users were Male.


- 

- 

- 



## limitation 
example
- We got missing data which we fix some of them .

- 
